src/hooks/use-search/use-search.ts

```diff
@@ -0,0 +1,247 @@
+import { Dispatch, useEffect, useReducer } from 'react';
+// @ts-ignore - addsearch has no types
+import AddSearchClient from 'addsearch-js-client';
+
+// NOTE: all the types are documented here https://www.addsearch.com/docs/api/search-results/
+type Image = {
+  main: string;
+  main_b64: string;
+  capture: string;
+};
+
+type DocumentType = 'html' | 'pdf' | 'doc' | 'docx' | 'ppt' | 'pptx';
+
+type Hit = {
+  id: string;
+  url: string;
+  title: string;
+  meta_description: string;
+  meta_categories: string[]; // <meta name="addsearch-category" content="features/api" />
+  custom_fields: Record<string, string> | Record<string, string[]>;
+  highlight: string; // NOTE: comes with html tags like <em>, etc.
+  ts: string; // NOTE: essentially an ISO date string
+  categories: string[];
+  document_type: DocumentType;
+  images: Image;
+  score: number; // NOTE: a number between 0 and 1
+};
+
+type Result = {
+  error?: { message: string } | null;
+  page: number;
+  total_hits: number;
+  processing_time_ms: number;
+  hits: Hit[];
+  facets: null;
+  rangeFacets: null;
+};
+
+export type State = {
+  client: string | null;
+  query?: string;
+  page?: Result['page'];
```

Comment:
page should be always set, by default to `1` - was the compiler saying otherwise?

Reply:
it does default to `1` but we don't always pass a page in our `search()` function so that's why it's optional. I just reused `State` type for our `search` action payload 
```
const searchAction = (dispatch: Dispatch<Action>) => (payload: Pick<State, 'query' | 'page'>) => dispatch({ type: 'search', payload });
```

I just didn't want to create another type, but I can if it's preferred. 👍 🙂

Reply:
right, yeah that's fine. I was thinking of https://github.com/ably/website/pull/4849/files#diff-b580f33881698284e8a407553ae825a97c41f2f21a19b2699b548a8501c496b1R58. I think it's to keep types non-polymorpic where possible :)